src/main/java/com/cloudogu/scm/tracemonitor/TraceStore.java

```diff
@@ -24,60 +24,80 @@
-import java.util.stream.Collectors;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.function.Supplier;
 
+@Singleton
 public class TraceStore {
 
   private static final String STORE_NAME = "trace-monitor";
-  private final ConfigurationEntryStoreFactory storeFactory;
+  private final DataStoreFactory storeFactory;
   private final GlobalConfigStore globalConfigStore;
+  private final Striped<ReadWriteLock> locks = Striped.readWriteLock(10);
 
   @Inject
-  public TraceStore(ConfigurationEntryStoreFactory storeFactory, GlobalConfigStore globalConfigStore) {
+  public TraceStore(DataStoreFactory storeFactory, GlobalConfigStore globalConfigStore) {
     this.storeFactory = storeFactory;
     this.globalConfigStore = globalConfigStore;
   }
 
   public Collection<SpanContext> getAll() {
     SecurityUtils.getSubject().checkPermission("traceMonitor:read");
-    DataStore<StoreEntry> store = createStore();
     List<SpanContext> spans = new ArrayList<>();
-    store.getAll().values().forEach(entry -> spans.addAll(entry.getSpans()));
+    doSynchronized("", false, () -> {
+      createStore()
+        .getAll()
+        .values()
+        .forEach(entry -> spans.addAll(entry.getSpans()));
+      return null;
+    });
     return spans;
   }
 
   public Collection<SpanContext> get(String kind) {
-    return getAll().stream()
-      .filter(span -> kind.equalsIgnoreCase(span.getKind()))
-      .collect(Collectors.toList());
-  }
+    SecurityUtils.getSubject().checkPermission("traceMonitor:read");
 
-  void add(SpanContext spanContext) {
-    DataStore<StoreEntry> store = createStore();
-    StoreEntry storeEntry = store.get(spanContext.getKind());
-    int configuredStoreSize = globalConfigStore.get().getStoreSize();
-    if (storeEntry == null) {
-      storeEntry = new StoreEntry(configuredStoreSize);
-    }
-    storeEntry = resizeStoreEntryMaxSize(storeEntry, configuredStoreSize);
-    storeEntry.getSpans().add(spanContext);
-    store.put(spanContext.getKind(), storeEntry);
+    final ArrayList<SpanContext> spans = new ArrayList<>();
+    doSynchronized(kind, false, () -> {
+      spans.addAll(createStore().get(kind).getSpans());
+      return null;
+    });
 
+    return spans;
+  }
+
+  synchronized void add(SpanContext spanContext) {
+    doSynchronized(spanContext.getKind(), true, () -> {
+      DataStore<StoreEntry> store = createStore();
+      StoreEntry storeEntry = store.get(spanContext.getKind());
+      int configuredStoreSize = globalConfigStore.get().getStoreSize();
+      if (storeEntry == null) {
+        storeEntry = new StoreEntry(configuredStoreSize);
+      }
+      storeEntry = resizeStoreEntryMaxSize(storeEntry, configuredStoreSize);
+      storeEntry.getSpans().add(spanContext);
+      store.put(spanContext.getKind(), storeEntry);
+      return null;
+    });
   }
 
   private StoreEntry resizeStoreEntryMaxSize(StoreEntry storeEntry, int configuredStoreSize) {
```

Comment:
This is for the line below: When the new size is bigger than the current one of the queue, you don't need to copy each value, you just would have to set the new size. But this is a minor issue (probably sizes will not be changed that often).